cpp/src/arrow/dataset/parquet_encryption_config.h

```diff
@@ -0,0 +1,70 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#pragma once
+
+#include "arrow/dataset/type_fwd.h"
+#include "parquet/encryption/crypto_factory.h"
+#include "parquet/encryption/encryption.h"
+#include "parquet/encryption/kms_client.h"
+
+namespace arrow {
+namespace dataset {
+
+/// core class, that translates the parameters of high level encryption
+struct ARROW_DS_EXPORT DatasetEncryptionConfiguration {
+  DatasetEncryptionConfiguration()
+      : kms_connection_config(
+            std::make_shared<parquet::encryption::KmsConnectionConfig>()) {}
+
+  void Setup(
+      std::shared_ptr<parquet::encryption::CryptoFactory> crypto_factory,
+      std::shared_ptr<parquet::encryption::KmsConnectionConfig> kms_connection_config,
+      std::shared_ptr<parquet::encryption::EncryptionConfiguration> encryption_config) {
+    this->crypto_factory = std::move(crypto_factory);
+    this->kms_connection_config = std::move(kms_connection_config);
+    this->encryption_config = std::move(encryption_config);
+  }
+
+  std::shared_ptr<parquet::encryption::CryptoFactory> crypto_factory;
+  std::shared_ptr<parquet::encryption::KmsConnectionConfig> kms_connection_config;
+  std::shared_ptr<parquet::encryption::EncryptionConfiguration> encryption_config;
+};
```

Comment:
```suggestion
struct ARROW_DS_EXPORT DatasetEncryptionConfiguration {
  std::shared_ptr<parquet::encryption::CryptoFactory> crypto_factory;
  std::shared_ptr<parquet::encryption::KmsConnectionConfig> kms_connection_config;
  std::shared_ptr<parquet::encryption::EncryptionConfiguration> encryption_config;
};
```

Have you considered to simply define it as the above? From you test cases, it seems that not all parameters are required. So `Setup` function is not that helpful if none of checks is done inside. By doing this, it enables aggregate initialization in C++20 like this:
```
auto conf = {
  .crypto_factory = xxx,
  .kms_connection_config = std::make_shared<parquet::encryption::KmsConnectionConfig>()
};
```

I am not saying we have to do this. Just a suggestion.

Reply:
Thanks for the explanation! Then I think we still can consolidate DatasetEncryptionConfiguration and DatasetDecryptionConfiguration into a single DatasetEncryptionConfiguration and do not provide default values. WDYT? Something like
```cpp
struct ARROW_DS_EXPORT DatasetEncryptionConfiguration {
  void Setup(
      std::shared_ptr<parquet::encryption::CryptoFactory> crypto_factory,
      std::shared_ptr<parquet::encryption::KmsConnectionConfig> kms_connection_config,
      std::shared_ptr<parquet::encryption::EncryptionConfiguration> encryption_config,
      std::shared_ptr<parquet::encryption::DecryptionConfiguration> decryption_config) {
    this->crypto_factory = std::move(crypto_factory);
    this->kms_connection_config = std::move(kms_connection_config);
    this->encryption_config = std::move(encryption_config);
    this->decryption_config = std::move(decryption_config);
  }

  std::shared_ptr<parquet::encryption::CryptoFactory> crypto_factory;
  std::shared_ptr<parquet::encryption::KmsConnectionConfig> kms_connection_config;
  std::shared_ptr<parquet::encryption::EncryptionConfiguration> encryption_config;
  std::shared_ptr<parquet::encryption::DecryptionConfiguration> decryption_config;
};
```

Reply:
If there is a minimum requirement for some fields to be not null, we should also check them in the Setup call.